samples/s3/s3-cp.c

```diff
@@ -233,42 +233,42 @@ int s_input_seek(struct aws_input_stream *stream, int64_t offset, enum aws_strea
         aws_string_destroy(state);
     }
 
-    return update_stream->wrapped_stream->vtable->seek(stream, offset, basis);
+    return aws_input_stream_seek(update_stream->wrapped_stream, offset, basis);
 }
 
 int s_input_read(struct aws_input_stream *stream, struct aws_byte_buf *dest) {
-    struct progress_update_stream *update_stream = stream->impl;
+    struct progress_update_stream *update_stream = aws_input_stream_get_impl(stream);
 
     size_t current_len = dest->len;
-    int val = update_stream->wrapped_stream->vtable->read(update_stream->wrapped_stream, dest);
+    int val = aws_input_stream_read(update_stream->wrapped_stream, dest);
     size_t progress = dest->len - current_len;
     progress_listener_update_progress(update_stream->transfer->listener, progress);
     return val;
 }
 
 static int s_input_get_status(struct aws_input_stream *stream, struct aws_stream_status *status) {
-    struct progress_update_stream *update_stream = stream->impl;
+    struct progress_update_stream *update_stream = aws_input_stream_get_impl(stream);
 
-    return update_stream->wrapped_stream->vtable->get_status(update_stream->wrapped_stream, status);
+    return aws_input_stream_get_status(update_stream->wrapped_stream, status);
 }
 
 static int s_input_get_length(struct aws_input_stream *stream, int64_t *out_length) {
-    struct progress_update_stream *update_stream = stream->impl;
-    return update_stream->wrapped_stream->vtable->get_length(update_stream->wrapped_stream, out_length);
+    struct progress_update_stream *update_stream = aws_input_stream_get_impl(stream);
+    return aws_input_stream_get_length(update_stream->wrapped_stream, out_length);
 }
 
 static void s_input_destroy(struct aws_input_stream *stream) {
-    struct progress_update_stream *update_stream = stream->impl;
-    update_stream->wrapped_stream->vtable->destroy(update_stream->wrapped_stream);
-    aws_mem_release(stream->allocator, update_stream);
+    struct progress_update_stream *update_stream = aws_input_stream_get_impl(stream);
+    aws_input_stream_release(update_stream->wrapped_stream);
+    aws_mem_release(update_stream->allocator, update_stream);
 }
 
 static struct aws_input_stream_vtable s_update_input_stream_vtable = {
     .get_length = s_input_get_length,
     .seek = s_input_seek,
     .read = s_input_read,
     .get_status = s_input_get_status,
-    .destroy = s_input_destroy,
+    .impl_destroy = s_input_destroy,
```

Comment:
Are we renaming `.destroy` to `.impl_destroy`, or keeping both? Not sure if the `impl_` prefix makes it more readable from the interface perspective.

Reply:
we are renaming it. So, for now the input stream will be clean up when the refcount drops to zero, and when it happens the `impl_destroy` will be invoked to clean up the impl passed in.